test/functional/wallet_coinselection.py

```diff
@@ -0,0 +1,261 @@
+
+        # 3) target is too far away for BnB to find a solution
+        # effective value of the big coin is smaller than target + MIN_CHANGE
+        #  * BnB can't find a solution as the distance from target is farther than cost of change
+        #    6061550 + 5000000*19-(42+148*20)*350-99999319 = 11531
+        #    cost_of_change = 11530 = 31*350 + 68*10
+        #  * knapsack finds closest solution to target + MIN_CHANGE with 21 small coins
+        #  * SRD still occasionally find better solution
+        fee_total, max_fee = select_coins(big_coin=1.01031, target=0.99999319, iterations=100)
+        srd_fee = (41 + 147*21 + 31) * fee_rate / COIN
+        assert_approx(max_fee, srd_fee)
+        assert_greater_than(srd_fee * 100, fee_total)
+        assert_greater_than(fee_total, best_fee * 100)
+
+        # 4) knapsack latches to exact match + MIN_CHANGE
+        # SRD still occasionally find better solution
+        fee_total, max_fee = select_coins(big_coin=1.00051, target=0.99, iterations=100)
+        knapsack_fee = (41 + 31 + 147*20) * fee_rate / COIN
+        assert_approx(max_fee, knapsack_fee)
+        assert_greater_than(knapsack_fee * 100, fee_total)
+        assert_greater_than(fee_total, best_fee * 100)
+
+    def test_one_big_coin(self):
+        w0 = self.nodes[0].get_wallet_rpc(self.default_wallet_name)
+        self.nodes[0].createwallet(wallet_name="test_one_big_coin")
+        w1 = self.nodes[0].get_wallet_rpc("test_one_big_coin")
+
+        w0.sendtoaddress(w1.getnewaddress("", "legacy"), 50)
+        self.generate(self.nodes[0], 1)
+
+        # change is less than MIN_CHANGE but we still create it as we don't have other options
+        result = w1.fundrawtransaction(w1.createrawtransaction([], {w0.getnewaddress(): 49.9999}), {"fee_rate": 10})
+        tx = w1.decoderawtransaction(result['hex'])
+        assert_equal(len(tx['vin']), 1)
+        assert_equal(len(tx['vout']), 2)
+
+    def test_one_big_and_small_coin(self):
+        w0 = self.nodes[0].get_wallet_rpc(self.default_wallet_name)
+        self.nodes[0].createwallet(wallet_name="test_one_big_and_small_coin")
+        w1 = self.nodes[0].get_wallet_rpc("test_one_big_and_small_coin")
+
+        w0.sendtoaddress(w1.getnewaddress("", "legacy"), 0.011)
+        w0.sendtoaddress(w1.getnewaddress("", "legacy"), 50)
+        self.generate(self.nodes[0], 1)
+
+        dummyaddress = w0.getnewaddress()
+        fee_rate = 20
+
+        def select_coins_for(target):
+            result = w1.fundrawtransaction(w1.createrawtransaction([], {dummyaddress: target}), {"fee_rate": fee_rate})
+            return w1.decoderawtransaction(result['hex'])
+
+        # if change is > MIN_CHANGE than we just spend big coin and create a change
+        tx = select_coins_for(49.9)
+        assert_equal(len(tx['vin']), 1)
+        assert_equal(len(tx['vout']), 2)
+
+        # if small coins > MIN_CHANGE than we just add one small coin to the big one and create change
```

Comment:
This comment confuses me. After paying the fees, the big coin doesn't have enough funds to pay fees and create the output here, so we would always need both coins.

Reply:
Oops. The idea was that you have more than 1 small coin. If small coins > MIN_CHANGE that the tests in `test_one_big_and_many_small_coins` doesn't work. Any idea how to improve that? For now, added more than one small coin in this test case